src/plugins/intel_gpu/src/graph/impls/ocl/reorder.cpp

```diff
@@ -108,12 +110,54 @@ struct reorder_impl : typed_primitive_impl_ocl<reorder> {
         auto kernel_params = get_kernel_params(impl_param, true);
         (_kernel_data.update_dispatch_data_func)(kernel_params.first, _kernel_data);
     }
+
+    static std::unique_ptr<primitive_impl> create(const reorder_node& arg, const kernel_impl_params& impl_param) {
+        bool is_reorder_weights = format::is_weights_format(impl_param.get_input_layout().format) ||
+                                  format::is_weights_format(impl_param.get_output_layout().format);
+        if (!is_reorder_weights) {
+            if (arg.can_be_optimized()) {
+                return make_unique<reorder_impl>(kernel_selector::kernel_data{});
+            }
+            auto kernel_params = reorder_impl::get_kernel_params(reorder_impl::static_canonicalize_shapes(impl_param));
+            kernel_params.first.is_shape_agnostic = impl_param.is_dynamic();
+            kernel_params.first.set_dynamic_shape_offsets();
+            auto& kernel_selector = reorder_impl::kernel_selector_t::Instance();
+            auto best_kernel = kernel_selector.get_best_kernel(kernel_params.first, kernel_params.second);
+
+            return make_unique<reorder_impl>(best_kernel);
+        } else {
+            return create_reorder_weigths(impl_param);
+        }
+    }
+
+    static std::unique_ptr<primitive_impl> create_reorder_weigths(const kernel_impl_params& impl_param) {
+        const auto& prim = impl_param.typed_desc<reorder>();
+        const auto& weights_params = prim->weights_reorder_params;
+        auto& kernel_selector = kernel_selector::ReorderWeightsKernelSelector::Instance();
+
+        OPENVINO_ASSERT(impl_param.get_input_layout().bytes_count() == weights_params->get_input_layout().bytes_count(),
+                        "[GPU] Input layout doesn't match required reorder weights layout");
```

Comment:
Because layouts can have differences in some cases (for example: after fake reshape weights for fc, info about this is not available for constant nodes during execution propagate_constants pass, therefore in impl_param will be stored another shape)